agent/ecs_meta_data.go

```diff
@@ -13,20 +14,50 @@ type ECSMetadata struct {
 func (e ECSMetadata) Get() (map[string]string, error) {
 	metaData := make(map[string]string)
 
-	ecsMeta, err := metadata.GetContainer(context.Background(), &http.Client{})
+	taskMeta, err := metadata.GetTask(context.Background(), &http.Client{})
 	if err != nil {
 		return metaData, err
 	}
 
-	switch m := ecsMeta.(type) {
+	switch m := taskMeta.(type) {
+	case *metadata.TaskMetadataV3:
+		metaData["ecs:task-arn"] = m.TaskARN
+		if m.Limits.CPU != 0 {
+			metaData["ecs:cpu-limit"] = strconv.FormatFloat(m.Limits.CPU, 'f', -1, 64)
+		}
+		if m.Limits.Memory != 0 {
+			metaData["ecs:memory-limit"] = strconv.Itoa(m.Limits.Memory)
+		}
+	case *metadata.TaskMetadataV4:
+		// This might be missing on some versions of Fargate which
+		// seems to unmarshal as "true"
+		if m.AvailabilityZone != "true" {
+			metaData["ecs:availability-zone"] = m.AvailabilityZone
+		}
```

Comment:
it's weird that this is required. In my manual testing (and according to the docs) Fargate 1.4.0 (using v4 metadata) always has the AZ, and Fargate 1.3.0 (using v3 metadata) doesn't.

We're getting `ecs:availability-zone=true` with Fargate 1.3.0, but we're only setting that tag in the v4 metadata code path. I'd have thought this code path wouldn't run in that case 🤔

Reply:
Oh, I see what's happening. I ran this on a Fargate 1.3.0 task:

```
#!/bin/sh

echo "ECS_CONTAINER_METADATA_URI_V4:$ECS_CONTAINER_METADATA_URI_V4"
echo "ECS_CONTAINER_METADATA_URI:$ECS_CONTAINER_METADATA_URI"
echo
echo


echo
echo "Task meta-data v4:"
curl "$ECS_CONTAINER_METADATA_URI_V4/task"


echo
echo "Task meta-data v3:"
curl "$ECS_CONTAINER_METADATA_URI/task"
```

... and I get this:

```
$ ./test.sh
ECS_CONTAINER_METADATA_URI_V4:http://169.254.170.2/v4/938a7299-e970-4ebd-b458-fc9dcac665b6
ECS_CONTAINER_METADATA_URI:http://169.254.170.2/v3/938a7299-e970-4ebd-b458-fc9dcac665b6
 
 
Task meta-data v4:
{...} # v4 shape, but no `AvailabilityZone` key. probably other differences too
Task meta-data v3:
{...}
```

So on a Fargate 1.3.0 task we still get a v4 metadata URL and https://github.com/BrunoScheufler/aws-ecs-metadata-go is using that, so we end up in the `case *metadata.TaskMetadataV4` code path. 

I'd suggest just removing the v3 code path in this file, but I'm not sure if the behaviour I observed will be the same in other cases (like `ECS/EC2`). Given how rare fargate 1.3.0 is likely to be, the check to skip a value of `true` is probably good enough.

Reply:
Yeah that all feels like a bit of a mess. I think ignoring "true" is Good Enough:tm:

Reply:
agree.

I'm tempted to parse the `TaskARN` value (which seems available in all ECS environments) to extract the region reliably. I don't want to hold up this PR though, so if I want to consider that I'll wait for this to merge and open a follow up PR

Reply:
Great minds: https://3.basecamp.com/3453178/buckets/27608512/chats/4937326106@5910377876